site/docs/hosting/vps.md

```diff
@@ -45,131 +137,638 @@ cd /etc/systemd/system
+```
+
+For the bot to work, make the configuration look like this:
+
+```text:no-line-numbers
+<domain> {
+  root * /usr/share/caddy
+  file_server
+  reverse_proxy /<token> localhost:8000
+}
+```
+
+replacing `<domain>` with your domain and `<token>` with your bot token.
+
+Reload Caddy every time as a site configuration file using the following command:
+
+```sh:no-line-numbers
+systemctl reload caddy
+```
+
+Now all requests to the address `https://<domain>/<token>` will be redirected to the address `http://localhost:8000/<token>`, where the bot's webhook is running.
+
+#### Connecting a Webhook to Telegram
+
+All you have to do is tell Telegram where to send the updates.
+To do this, open your browser and visit the page at the following link:
+
+```text:no-line-numbers
+https://api.telegram.org/bot<token>/setWebhook?url=https://<domain>/<token>
 ```
 
-### Advanced Information
+replacing `<token>` with your bot token and `<domain>` with your domain.
+
+## CI/CD
+
+[CI/CD](https://about.gitlab.com/topics/ci-cd/) is an important part of the modern software development process.
+This guide covers almost all of the [CI/CD pipeline](https://about.gitlab.com/topics/ci-cd/cicd-pipeline/).
+
+We will focus on writing scripts for GitHub and GitLab.
+You can easily adapt the examples below to your CI/CD service of choice, such as Jenkins, Buddy, etc., if needed.
+
+:::tip Self-hosted runner
+GitHub and GitLab offer a certain amount of resources for free to complete your tasks.
+However, when setting up your pipeline, you may quickly use them all up, which will require you to pay money for additional resources or the tasks will not run.
+To prevent this from happening, we recommend installing a self-hosted runner on your computer so that tasks run on your computer.
+This way, you'll get rid of the limits and be able to run pipelines of any complexity and capacity (almost).
+:::
+
+### SSH Keys
+
+To deliver files to the server, you need to set up passwordless authentication, which is implemented using SSH keys.
+
+The following commands should be run on your personal computer.
+
+Change to the directory with the SSH keys:
+
+```sh:no-line-numbers
+cd ~/.ssh
+```
+
+Generate a new key pair:
+
+<CodeGroup>
+  <CodeGroupItem title="GitHub" active>
+
+```sh:no-line-numbers
+ssh-keygen -t rsa -m PEM
+```
+
+</CodeGroupItem>
+  <CodeGroupItem title="GitLab">
+
+```sh:no-line-numbers
+ssh-keygen -t ed25519
+```
+
+</CodeGroupItem>
+</CodeGroup>
+
+This command will generate a public and private key of the type and format you want for GitHub and GitLab.
+You can also specify a custom key name if you wish.
+Note that you should skip the fingerprint generation step (just press `Enter`).
+
+Next, send the **public** key to the server:
+
+```sh:no-line-numbers
+ssh-copy-id -i <key-name>.pub root@<ip-address>
+```
+
+replacing `<key-name>` with the name of the generated key and `<ip-address>` with the IP address of your server.
```

Comment:
```suggestion
Replace `<key-name>` with the name of the generated key and `<ip-address>` with the IP address of your server.
```

Reply:
Why did you decide to write this way? I write like this everywhere else. Should I make a similar change elsewhere?

Reply:
_replacing_ requires a sentence before that: _Do X, replacing X by Y._ I don't find it very intuitive that a code snippet is supposed to represent this part of the sentence. The sentence before the code snippet is separated by a colon (:), so it can't take that role, either.

Besides, if an English sentence can be split up into two sentences, it's usually preferable to do so. English isn't meant to be used for complex sentences. It lacks the grammatical features of your and mine mother tongue :(

Reply:
So yes, it would make sense to make similar changes elsewhere, but Roj and me are going to do a final pass over this and nitpick all the little style and language things, so we're gonna catch these cases anyway. You don't _have to_ fix them now.

Reply:
I see. I wrote _replacing_ because that's how it was written in other articles about hosting.